core/src/main/scala/com/metamx/tranquility/tranquilizer/Tranquilizer.scala

```diff
@@ -258,14 +254,15 @@ class Tranquilizer[MessageType] private(
 
     propagate respond { result =>
       result match {
-        case Return(n) if n == myBuffer.size =>
-          log.debug(s"Sent $n out of ${myBuffer.size} messages. Marking batch complete.")
-          myBuffer.foreach(_.future.setValue(()))
-
-        case Return(n) =>
-          log.debug(s"Sent $n out of ${myBuffer.size} messages. Marking batch dropped.")
-          val e = new MessageDroppedException(s"Batch not complete ($n/${myBuffer.size})")
-          myBuffer.foreach(_.future.setException(e))
+        case Return(bitset) =>
+          log.debug(s"Sent ${bitset.size} out of ${myBuffer.size} messages.")
+          for ((holder, index) <- Beam.index(myBuffer)) {
+            if (bitset.contains(index)) {
+              holder.future.setValue(())
+            } else {
+              holder.future.setException(MessageDroppedException.instance)
```

Comment:
Also the message is less useful now, since the exception to be applied to *all* messages (even the successful ones) and just tell you in the message how many were successful. That was not really that great. Now the exception is only applied to the actually dropped messages.